File: 2024-11-15/try_default_arg.rb
#
# Timeオブジェクトを文字列に変換する
# デフォルト引数の初期化タイミングを確認するために書いている
#
def get_time_string(time = Time.now)
    time.to_s
end

ONE_DAY_SECS = 24 * 60 * 60

if __FILE__ == $PROGRAM_NAME
  p $PROGRAM_NAME
  p get_time_string
  p get_time_string(Time.now - ONE_DAY_SECS) # -> 引数に指定された通りの日になる
  p get_time_string(Time.now - 2 *  ONE_DAY_SECS) # -> 引数に指定された通りの日になる
  sleep(3)
  p get_time_string # -> 3s後になる。つまり、デフォルト引数の値は都度評価される
end

File: 2024-11-15/try_default_arg.py
from datetime import datetime, timedelta
from time import sleep


def get_time_str(time=datetime.now()):
    return str(time)


if __name__ == "__main__":
    print(f"{get_time_str()}")
    print(f"{get_time_str(datetime.now() - timedelta(days=1))}")
    print(f"{get_time_str(datetime.now() - timedelta(days=2))}")
    sleep(3)
    print(
        f"{get_time_str()}"
    )  # -> 最初に呼び出したときと同じ。関数が評価されたときに値も評価されて、その一回のみ
